File: original_app/GregoireConverter.gabc/NeumeFormat.cs
using System;
using System.Text;

namespace GregoireConverter.gabc
{
	public class NeumeFormat
	{
		public int SizeX;
		public int SizeY;
		public NeumePart[] Parts;

		public NeumeFormat ()
		{
		}

		public NeumeFormat (int sizeX, int sizeY, params NeumePart[] parts)
		{
			SizeX = sizeX;
			SizeY = sizeY;
			Parts = parts;
		}

		public override string ToString()
		{
			var sb = new StringBuilder();
			foreach (var part in Parts)
			{
				sb.Append(part.Before);
				sb.Append(part.Height);
				sb.Append(part.After);
			}
			return sb.ToString();
		}
	}
}

File: web/grg2gabc/src/app/core/gabc/neume-format.ts
import { NeumePart } from './neume-part';

export class NeumeFormat {
  sizeX: number;
  sizeY: number;
  parts: NeumePart[];

  constructor(sizeX: number, sizeY: number, ...parts: NeumePart[]) {
    this.sizeX = sizeX;
    this.sizeY = sizeY;
    this.parts = parts;
  }

  toString(): string {
    return this.parts.map(p => p.before + p.height + p.after).join('');
  }

  clone(): NeumeFormat {
    return new NeumeFormat(
      this.sizeX,
      this.sizeY,
      ...this.parts.map(p => new NeumePart(p.startX, p.startY, p.before, p.height, p.after, p.sizeX, p.sizeY))
    );
  }
}
